src/newgrf.cpp

```diff
@@ -653,8 +653,9 @@ static Engine *GetNewEngine(const GRFFile *file, VehicleType type, uint16_t inte
 
 		/* Reserve the engine slot */
 		if (!static_access) {
-			EngineIDMapping &eid = _engine_mngr.mappings[engine];
-			eid.grfid = scope_grfid; // Note: this is INVALID_GRFID if dynamic_engines is disabled, so no reservation
+			/* Replace an existing slot, so remove the existing map entry for it. */
+			_engine_mngr.EraseID(type, engine);
+			_engine_mngr.SetID(type, internal_id, scope_grfid, std::min<uint8_t>(internal_id, _engine_counts[type]), e->index);
```

Comment:
This made me question whether this is suboptimal. Why wouldn't you just reuse the original location, but I reckon the primary reason this gets called is when it's replaced by an engine with a different grfid. Is it maybe worth to describe that?
'Replace an existing slot. Given the grfid (or internal_id) is likely different, the replacement will end up in a different location, so erase and add'

Reply:
Yup. A rotate between the old and new position would allow the slot to be moved, moving only entries between them without moving everything after. But...